sdm845-mainline: fix firmware issue, update to 6.4.0, and sync config with upstream

linyinfeng opened the pull request:
This pull request:

1. Contains <https://github.com/NixOS/mobile-nixos/pull/627>.
2. Fixes <https://github.com/NixOS/mobile-nixos/pull/627#issuecomment-1546920191> in <https://github.com/NixOS/mobile-nixos/commit/9619b07e29f092e6d11bd4f5c874f39b55554601>.
3. Because firmware in stage-2 is xz compressed (<https://github.com/linyinfeng/nixpkgs/blob/2a3797a3d549e4d59c33719fa9d5328aa6eae799/nixos/modules/services/hardware/udev.nix#L169>), enable FW_LOADER_COMPRESS_XZ in <https://github.com/NixOS/mobile-nixos/commit/dba7ed4cf0e0a4fc053e13d20e1be9ad0a35219c>.
4. Update sdm845-mainline kernel to upstream 6.3.4 (<https://gitlab.com/sdm845-mainline/linux/-/tags/sdm845-6.3.4>), then sync config with postmarketOS (https://gitlab.com/postmarketOS/pmaports/-/blob/016e512c830f0dffbfad2b5ed4c90da50bba9120/device/community/linux-postmarketos-qcom-sdm845/config-postmarketos-qcom-sdm845.aarch64).
5. Update sdm845-mainline kernel to upstream 6.4.0 (<https://gitlab.com/sdm845-mainline/linux/-/tags/sdm845-6.4-r1>)

Test on my enchilada (OnePlus 6) with gnome desktop (`github:linyinfeng/dotfiles#enchilada`):

Working: display, touchscreen, WiFi (wrong link speed reported), 4G mobile data, speaker and eadphone jack, gadgetfs (manually configured RNDIS works in stage 2), battery and charging, suspend, SMS.

Not working: sensors, microphone, camera, and call (connected but no sound)~, and SMS~.

Not tested: FDE, OTG, GPS, NFC, 3D Acceleration, fast charging (I don't have VOOC charger), etc.

![enchilada-6 3 4-gnome](https://github.com/NixOS/mobile-nixos/assets/11229748/93b9c504-61b7-42e7-917f-8ce46e9fac3a)

Some issues:
1. It seems that stage-1 init crashes with `mobile.boot.stage-1.networking.enable = true` or `mobile.adbd.enable = true`. Turn these gadgetfs-related options off then stage-1 works well.
2. `iio-sensor-proxy.service` stopped immediately after start.
   ```
   Jun 18 13:00:00 enchilada systemd[1]: Starting IIO Sensor Proxy service...
   Jun 18 13:00:00 enchilada systemd[1]: Started IIO Sensor Proxy service.
   Jun 18 13:00:00 enchilada systemd[1]: iio-sensor-proxy.service: Deactivated successfully.
   ```
3. Wired upstream GNOME issues like <https://gitlab.gnome.org/GNOME/mutter/-/issues/2836>. Connect a mouse before dragging any window using the touchscreen to workaround this issue.

samueldr committed:
kernel/builder: Ignore all CONFIG_INIT_STACK options

These are influenced by toolchain versions. As such, it is not useful to
be pedantic about them.

linyinfeng committed:
sdm845-mainline: Add extra a630 firmware to stage-1

linyinfeng committed:
sdm845-mainline/kernel: update to 6.4.0

samueldr commented:
Hi! Thanks for the update!

I'll try to make time to test quickly on enchilada.

* * *

> Not working: sensors, microphone, camera, call (connected but no sound), and SMS.

It would be helpful to state whether or not those *should* be working, or those *worked in the past*.

In other words, to know if there are regressions or not.

> Not tested: FDE, OTG, GPS, NFC, 3D Acceleration, fast charging (I don't have VOOC charger), etc.

Of those:

 - FDE is mostly out-of-scope for hardware specifics... If touch and display works, and the kernel options haven't been disabled, it'll continue working just fine.
 - 3D accel *should be considered working* given you're using (I suspect) gnome desktop on wayland.
 - OTG switching is known from upstream to require driver work for sensing.

MatthewCroughan commented:
In the past I have had SMS working, and sound via the speakers, though the microphone was very temperamental. Audio work is being done in sdm845-mainline some devices should be working better, and maybe the userspace software configuration needs to change as a result https://gitlab.com/sdm845-mainline/linux/-/commit/562d218d632dc87d7065bdd92a171bc24d716d4f, a lot of work has been going into the `sdm845-shift-axolotl` that has not yet been done for the `oneplus-common`.

I'm going to test this PR now, and report more.

MatthewCroughan commented:
I have been using this for the past few days, and I have been having great success with it. Suspend works, No new issues with this kernel as far as I can tell, on both fajita and enchilada, though I have not been able to test gps, nfc or modem as I don't make use of them at the moment.

chuangzhu reviewed (approved):
Works for me. Previously broken USB gadget is also fixed with this.

MatthewCroughan commented:
Tested SMS and mobile networking just now, it works.

linyinfeng commented:
> > Not working: sensors, microphone, camera, call (connected but no sound), and SMS.
> 
> SMS works for me, though. I use `chatty` and have `programs.calls.enable = true`.
> 
> > Wired upstream GNOME issues like https://gitlab.gnome.org/GNOME/mutter/-/issues/2836. Connect a mouse before dragging any window using the touchscreen to workaround this issue.
> 
> FYI, I have a working overlay for GNOME Mobile here: https://github.com/chuangzhu/nixpkgs-gnome-mobile

Seems that my userspace configuration is wrong causing SMS not working. I tried Plasma mobile and SMS works. I will correct the test result in the description. 

Thank you for the overlay! I will have a try.

samueldr committed:
sdm845-mainline: Normalize kernel config for 6.4.0

The normalization was actually made from the state of the config file on
commit 9f0289b65b2e2a430f8f9c4a52e3b2a6e99758f2.

Co-authored-by: Lin Yinfeng <lin.yinfeng@outlook.com>

samueldr commented:
This force push:

 - rebases on top of development
 - squashes (now) needless steps in the commit history
 - re-normalizes

I have verified it to boot without issues on my device.

Thank you @linyinfeng, and sorry it took so long to deal with!
